### PR TITLE
Add function to remap categorical labelings

### LINKED ISSUE #526: Add label matching function
Want a function to take in a true and predicted labeling, and reorder the labels of the predicted to match those of the true as well as possible. 

I have very simple code for this.

Will be useful for clustering tutorials at a bare minimum, but I also use this a lot when clustering in general.

### PR DESCRIPTION
**Pull request by bdpedigo**
#### Reference Issues/PRs
Fixes #526 

#### What does this implement/fix? Explain your changes.
Adds a function to try to remap labelings from one categorical labeling vector based on another. E.g. if one is clustering and you want the labels of predicted clusters to match some known truth. Has come up several times for clustering tutorials (in progress) and I also use this a decent amount in my own work so I think we should have it.

Here's a notebook with more explanation and a demo https://docs.neurodata.io/notebooks/pedigo/graspologic/clustering/2020/10/12/remapping-labels.html

#### Any other comments?
Would make more sense to put in sklearn someday, if they want it.

### CODE REVIEW

**Comment by bdpedigo** [comment]
@j1c or @asaadeldin11 could you have a look for JHU side?
@Nyecarr or @dwaynepryce or @bryantower could one of you have a look for MSR side?

build-documentation breaking looks to be something in dev since I've seen in several open PRs, looks like something with intersphinx breaking. microsoft checks also stalled on this PR.

File: graspologic/utils/utils.py [hunk 1]
```diff
@@ -701,3 +705,51 @@ def plug_in_variance_estimator(x):
         return covariances
 
     return plug_in_variance_estimator
+
+
+def remap_labels(y_true, y_pred, return_map=False):
+    """
+    Attempts to remap a categorical labeling (such as one predicted by a clustering
```
Comment (dwaynepryce) [inline graspologic/utils/utils.py:712]: I would reword this - attempts is too nebulous and can imply a variety of things you may not intend.  If you do intend attempts, I'd describe the optimal circumstance as well as an example or examples of the negative cases too where we didn't succeed in automatically relabeling.

File: graspologic/utils/utils.py [hunk 2]
```diff
@@ -8,7 +8,11 @@
 
 import networkx as nx
 import numpy as np
+from scipy.optimize import linear_sum_assignment
+from sklearn.metrics import confusion_matrix
+from sklearn.metrics._classification import _check_targets
```
Comment (dwaynepryce) [inline graspologic/utils/utils.py:13]: Since this is a private function in sklearn, we should copy it out into our own code base or rewrite it entirely (depending on licensing details).  We can't guarantee they'll maintain this and relying/using it in our own code will make us more fragile.

File: graspologic/utils/utils.py [hunk 3]
```diff
@@ -701,3 +705,51 @@ def plug_in_variance_estimator(x):
         return covariances
 
     return plug_in_variance_estimator
+
+
+def remap_labels(y_true, y_pred, return_map=False):
```
Comment (j1c) [inline graspologic/utils/utils.py:710]: Add type hints. This should be pretty simple for this function.

See [this](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html) for a quick guide.

File: graspologic/utils/utils.py (hunk 1 above)
Comment (bdpedigo) [reply graspologic/utils/utils.py]: Addressed. I expanded a little bit at the top and added a `Notes` section. If you want more than what is in the notes currently, I think it'd be easier at that point to just make an actual tutorial. I thought that was overkill for this function but I do already have a notebook linked at the top.

File: graspologic/utils/utils.py (hunk 2 above)
Comment (bdpedigo) [reply graspologic/utils/utils.py:13]: fair. their license is BSD-3 ([link](https://github.com/scikit-learn/scikit-learn/blob/master/COPYING)). Do you know the answer for that license? If not, I am not sure how to find out with legal.

File: graspologic/utils/utils.py (hunk 3 above)
Comment (bdpedigo) [reply graspologic/utils/utils.py]: done. though I need to check mypy output. do you know how to do typing for variable output possibilities? I looked around and couldn't find anything. not sure if it's important, I think the main use case is just getting the array back in this PR. 

also, this code should technically work on anything sklearn considers "array-like" which I assume is a massive list of stuff. For now I just put list, array, series since again, I think this is 99.9% of use cases

**Comment by bdpedigo** [comment]
@dwaynepryce @j1c any further comments?

File: graspologic/utils/utils.py (hunk 2 above)
Comment (bdpedigo) [reply graspologic/utils/utils.py:15]: @dwaynepryce or @carolyncb can you advise on this point?

Comment (bdpedigo) [reply graspologic/utils/utils.py:15]: update: just need to include in license file -@carolyncb

Comment (dwaynepryce) [reply graspologic/utils/utils.py:15]: @carolyncb - still think we need to run this by CELA, since we're inlining their code.  pretty sure we just have to maintain a header in the file, but what does this do to the overarching LICENSE.txt for the project? we should just get guidance.

File: graspologic/utils/utils.py (hunk 3 above)
Comment (dwaynepryce) [reply graspologic/utils/utils.py]: Union[type1, type2, type3]

Comment (dwaynepryce) [reply graspologic/utils/utils.py]: https://docs.python.org/3/library/typing.html#typing.Union

Comment (bdpedigo) [reply graspologic/utils/utils.py]: @dwaynepryce i was just talking about what stuff goes in the Union[]

Comment (bdpedigo) [reply graspologic/utils/utils.py]: since "array-like" is vague and I'm sure a lot of stuff would actually work

**Comment by bdpedigo** [comment]
@asaadeldin11 @j1c @dwaynepryce would one of you have time to review this, I think it is all done as far as I know. I've addressed:
- typing
- removed dependency on sklearn private code

would like to get this in as #491 #516 #539 all need it, I think.

File: graspologic/utils/utils.py [hunk 4]
```diff
@@ -701,3 +706,90 @@ def plug_in_variance_estimator(x):
...
+def remap_labels(
+    y_true: Union[List, np.ndarray, pd.Series],
+    y_pred: Union[List, np.ndarray, pd.Series],
+    return_map: bool = False,
+) -> np.ndarray:
+    """
+    Remaps a categorical labeling (such as one predicted by a clustering algorithm) to
+    match the labels used by another similar labeling.
+
+    Given two $n$-length vectors describing a categorical labeling of $n$ samples, this
```
Comment (bdpedigo) [inline graspologic/utils/utils.py:720]: I will fix the $$ and make math

**Comment by asaadeldin11** [comment]
i can review!

**Comment by bdpedigo** [comment]
@asaadeldin11 thanks, should be good to go whenever you are

File: graspologic/utils/utils.py [hunk 5]
```diff
@@ -701,3 +706,89 @@ def plug_in_variance_estimator(x):
...
+    y_pred: Union[List, np.ndarray, pd.Series],
+    return_map: bool = False,
+) -> np.ndarray:
+    """
+    Remaps a categorical labeling (such as one predicted by a clustering algorithm) to
+    match the labels used by another similar labeling.
+
+    Given two :math:`n`-length vectors describing a categorical labeling of :math:`n`
+    samples, this method reorders the labels of the second vector (`y_pred`) so that as
+    many samplesas possible from the two label vectors are in the same category.
```
Comment (asaadeldin11) [inline graspologic/utils/utils.py:722]:
```suggestion
    many samples as possible from the two label vectors are in the same category.
```

File: graspologic/utils/utils.py [hunk 6]
```diff
@@ -701,3 +706,89 @@ def plug_in_variance_estimator(x):
...
+    the categories in `y_pred` match that original category from `y_true`.
+    """
+    check_consistent_length(y_true, y_pred)
+    true_type = type_of_target(y_true)
+    pred_type = type_of_target(y_pred)
+
+    valid_target_types = {"binary", "multiclass"}
+    if (true_type not in valid_target_types) or (pred_type not in valid_target_types):
+        msg = "Elements of `y_true` and `y_pred` must represent a valid binary or "
+        msg += "multiclass labeling, see "
```
Comment (asaadeldin11) [inline graspologic/utils/utils.py:774]:
```suggestion
        msg = ("Elements of `y_true` and `y_pred` must represent a valid binary or "
              "multiclass labeling, see ")
```
Believe the multiline string can also be done like this. Both achieve the same thing, so whether or not you want to change is up to you.

**Review by asaadeldin11 — APPROVED** [review]
Few minor comments, overall looks good.
